### PR TITLE
add route directive task

### PR DESCRIPTION
**Pull request by kyessenov**
Needs https://github.com/istio/istio/pull/12505
Fixes https://github.com/istio/istio.io/issues/3673

/assign @douglas-reid 
/assign @geeknoid 
/assign @bianpengyuan 

Signed-off-by: Kuat Yessenov <kuat@google.com>

### CODE REVIEW

**Comment by geeknoid** [comment]
Deploy preview for *preliminary-istio* failed.

Built with commit 3caf2eec5a1946c8da99cffa465963c1d8ae7613

https://app.netlify.com/sites/preliminary-istio/deploys/5c8b206f897ef70009e8b67d

**Comment by geeknoid** [comment]
/hold

Given the test adapter is not part of the 1.1 sources, folks using 1.1 won't be able to try out this task (and some of the links in this doc are broken since the sources aren't in the istio/istio repo, yet for some reason the linter isn't catching those.). 

Should we park this PR until the 1.2 timeframe?

File: content/about/notes/1.1/index.md [hunk 1]
```diff
@@ -113,7 +113,8 @@ a richer visualization experience. See the [Kiali task](/docs/tasks/telemetry/ki
     - Improved the protocol between Envoy and Mixer.
 
 - **Control Headers and Routing**. Added the option to create adapters to influence
-an incoming request's headers and routing.
+an incoming request's headers and routing. See [Control Headers and Routing task](/docs/tasks/policy-enforcement/control-headers)
```
Comment (geeknoid) [inline content/about/notes/1.1/index.md]: See -> See the
Put the word "task" outside the link.

File: content/docs/tasks/policy-enforcement/control-headers/index.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+            prefix: /status
+        route:
+        - destination:
+            port:
+              number: 8000
+            host: httpbin
+    EOF
+    {{< /text >}}
+
+## Output-producing check adapters
```
Comment (geeknoid) [inline content/docs/tasks/policy-enforcement/control-headers/index.md]: remove the word "check", we don't use that with customers.

File: content/docs/tasks/policy-enforcement/control-headers/index.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+populate the output value, or return `NOT_FOUND` error status if the input
+instance key is not present in the lookup table.
+
+1. Deploy the demo adapter:
+
+    {{< text bash >}}
+    $ kubectl run keyval --image=gcr.io/istio-testing/keyval:release-1.1 --namespace istio-system --port 9070 --expose
+    {{< /text >}}
+
+1. Enable `keyval` out-of-process adapter in Mixer by deploying its template and configuration descriptors:
```
Comment (geeknoid) [inline content/docs/tasks/policy-enforcement/control-headers/index.md]: Enable the
Remove "out-of-process"
Remove "in Mixer"

File: content/docs/tasks/policy-enforcement/control-headers/index.md [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+      adapter: keyval
+      connection:
+        address: keyval:9070
+      params:
+        table:
+          jason: admin
+    EOF
+    {{< /text >}}
+
+1. Create an instance for the handler with `user` request header as a lookup key:
```
Comment (geeknoid) [inline content/docs/tasks/policy-enforcement/control-headers/index.md]: with the `user`

File: content/docs/tasks/policy-enforcement/control-headers/index.md [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    spec:
+      template: keyval
+      params:
+        key: request.headers["user"] | ""
+    EOF
+    {{< /text >}}
+
+## Request header operations
+
+1. Ensure that _httpbin_ service is accessible through the ingress gateway:
```
Comment (geeknoid) [inline content/docs/tasks/policy-enforcement/control-headers/index.md]: that -> the

File: content/docs/tasks/policy-enforcement/control-headers/index.md [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+      "headers": {
+        "Accept": "*/*",
+        "Content-Length": "0",
+        ...
+        "X-Envoy-Internal": "true"
+      }
+    }
+    {{< /text >}}
+
+   The output should be the request headers as they are received by _httpbin_ service.
```
Comment (geeknoid) [inline content/docs/tasks/policy-enforcement/control-headers/index.md]: received by the

File: content/docs/tasks/policy-enforcement/control-headers/index.md [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        "User": "jason",
+        "User-Agent": "curl/7.58.0",
+        "User-Group": "admin",
+        ...
+        "X-Envoy-Internal": "true"
+      }
+    }
+    {{< /text >}}
+
+   Note the presence of `user-group` header with the value derived from the
```
Comment (geeknoid) [inline content/docs/tasks/policy-enforcement/control-headers/index.md]: presence of the

File: content/docs/tasks/policy-enforcement/control-headers/index.md [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+   rule application of the adapter. The expression `x.output.value` in the rule
+   evaluates to the populated `value` field returned by `keyval` adapter.
```
Comment (geeknoid) [inline content/docs/tasks/policy-enforcement/control-headers/index.md]: by the

File: content/docs/tasks/policy-enforcement/control-headers/index.md [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+
+The adapter developers can also specify the client-facing error response if the policy
+check fails. Issue a request to the gateway with a non-existing user key:
+
+{{< text bash >}}
+$ curl -Huser:bill http://$INGRESS_HOST:$INGRESS_PORT/headers
+<error_detail>key "bill" not found</error_detail>
+{{< /text >}}
+
+The demo adapter uses a [custom error
```
Comment (geeknoid) [inline content/docs/tasks/policy-enforcement/control-headers/index.md]: Remove this paragraph and the sample Go code, this is not relevant to customers.

File: content/about/notes/1.1/index.md (hunk 1 above)
Comment (kyessenov) [reply content/about/notes/1.1/index.md]: Done

File: content/docs/tasks/policy-enforcement/control-headers/index.md (hunk 2 above)
Comment (kyessenov) [reply content/docs/tasks/policy-enforcement/control-headers/index.md]: Done

File: content/docs/tasks/policy-enforcement/control-headers/index.md (hunk 3 above)
Comment (kyessenov) [reply content/docs/tasks/policy-enforcement/control-headers/index.md]: Done

File: content/docs/tasks/policy-enforcement/control-headers/index.md (hunk 4 above)
Comment (kyessenov) [reply content/docs/tasks/policy-enforcement/control-headers/index.md]: Done

File: content/docs/tasks/policy-enforcement/control-headers/index.md (hunk 5 above)
Comment (kyessenov) [reply content/docs/tasks/policy-enforcement/control-headers/index.md]: Done

File: content/docs/tasks/policy-enforcement/control-headers/index.md (hunk 6 above)
Comment (kyessenov) [reply content/docs/tasks/policy-enforcement/control-headers/index.md]: Done

File: content/docs/tasks/policy-enforcement/control-headers/index.md (hunk 7 above)
Comment (kyessenov) [reply content/docs/tasks/policy-enforcement/control-headers/index.md]: Done

File: content/docs/tasks/policy-enforcement/control-headers/index.md (hunk 8 above)
Comment (kyessenov) [reply content/docs/tasks/policy-enforcement/control-headers/index.md]: Done

File: content/docs/tasks/policy-enforcement/control-headers/index.md (hunk 9 above)
Comment (kyessenov) [reply content/docs/tasks/policy-enforcement/control-headers/index.md]: Done

**Comment by kyessenov** [comment]
@geeknoid 

After chatting with Doug, we thought it's better to use a pre-compiled adapter that is very simple, and this `keyval` fits the need since it's extremely simplistic. The only dependency of this task is availability of the docker image (which I pushed manually) and the proto descriptors (that are best kept in samples since they are large).  We don't have any other public adapter suitable for this task. I pushed a PR to 1.1 with the descriptors https://github.com/istio/istio/pull/12524/files that we can hopefully get in before this task.

**Comment by geeknoid** [comment]
All righty, let's get the other PR pushed in istio/istio and then this one can go in.

**Comment by geeknoid** [comment]
The PR in istio/istio has been merged.

Could you try and rebase and repush this PR? I just want to make the links are working correctly (the link checker was previously turned off, I turned it on a few days ago)

**Comment by kyessenov** [comment]
Rebased and verified links work now

**Comment by kyessenov** [comment]
I don't know why circle job is stuck and have no way to retrigger it. I manually checked that those two links are accessible from the preview.

**Comment by geeknoid** [comment]
CircleCi is stuck on several jobs, it appears there's an outage.

**Pull request merged**